broadcast/broadcast.go

```diff
@@ -49,17 +56,49 @@ func NewBroadcaster(n int) *Broadcaster {
 	return &Broadcaster{capacity: n}
 }
 
-// Send broadcasts a message to the channel.
+// SendWithTimeout broadcasts a message to each listener's channel.
 // Sending on a closed channel causes a runtime panic.
-func (b *Broadcaster) Send(v interface{}) {
+// This method blocks for a duration of up to `timeout` on each channel.
+// Returns error(s) if it is unable to send on a given listener's channel within `timeout` duration.
+func (b *Broadcaster) SendWithTimeout(v interface{}, timeout time.Duration) error {
 	b.m.Lock()
 	defer b.m.Unlock()
 	if b.closed {
 		panic("broadcast: send after close")
 	}
-	for _, l := range b.listeners {
-		l <- v
+	var result *multierror.Error
+	for id, l := range b.listeners {
+		select {
+		case l <- v:
+			// Success!
+		case <-time.After(timeout):
+			err := fmt.Sprintf("unable to send to listener '%d'", id)
+			result = multierror.Append(result, errors.New(err))
```

Comment:
I wonder if the caller might prefer a cleaner way of identifying which listeners _failed_, so maybe they want to do a manual retry later. Maybe instead of a `errors.New`, using a defined error type with receivers `id` would be handier to manage.

Reply:
I thought about this, but then felt that since we were already returning a custom error type (multierror), this might be too many levels of custom errors? But happy to create 2nd level if this makes sense? In our own usage of this, it will likely mostly be a 'fire and forget' sort of usage, but maybe not always?

Reply:
Yeh the more I think about it, the more I think this is really a pub/sub broadcaster, so specific errors like this are maybe overkill... In fact, we don't actually provide a public API to get listeners by id (nor is the id publicly available via the listener), so the ID isn't really helpful in the end anyway 🤷‍♂ .

Reply:
We _could_ expose the ID on the listener... but again, this might be overkill.

Reply:
Yeah, I guess will depend on how Broadcaster is used... so maybe better not pre-decide and wait for a bit.

Reply:
Saying it differently, I agree to take the _avoid complexity_ route until more having more info.